### PR TITLE
feat(textures): parse KTX2 with BasisLoader

### PR DESCRIPTION
**Pull request by belom88**
We need to publish this PR to use these changes in the example application. Should we include this in `3.0-release` or publish the first beta-version?

### CODE REVIEW

File: modules/textures/docs/api-reference/compressed-texture-loader.md [hunk 1]
```diff
@@ -34,6 +34,6 @@ Returns an array of image data objects representing mip levels.
 
 ## Options
 
-| Option | Type | Default | Description |
-| ------ | ---- | ------- | ----------- |
-| N/A    |      |         |             |
+| Option                        | Type    | Default | Description                            |
+| ----------------------------- | ------- | ------- | -------------------------------------- |
+| `compressed-texture.useBasis` | Boolean | `false` | Use BasisLoader to decode KTX2 texture |
```
Comment (ibgreen) [inline modules/textures/docs/api-reference/compressed-texture-loader.md:39]: Thanks for documenting it. I am not sure we want this to be a public option, but good for now. We should audit the API before 3.1 release.

File: modules/textures/package.json [hunk 2]
```diff
@@ -42,6 +42,7 @@
     "build-crunch-worker": "webpack --entry ./src/workers/crunch-worker.ts --output ./dist/crunch-worker.js --config ../../scripts/webpack/worker.js"
   },
   "dependencies": {
+    "@loaders.gl/core": "3.0.1",
```
Comment (ibgreen) [inline modules/textures/package.json:45]: We need to break this dependency. Maybe create an issue for that.

File: modules/textures/src/basis-loader.ts [hunk 3]
```diff
@@ -18,7 +18,9 @@ export const BasisWorkerLoader = {
   options: {
     basis: {
       format: 'rgb565', // TODO: auto...
-      libraryPath: 'libs/'
+      libraryPath: 'libs/',
+      decoderFormat: 'Basis', // 'Basis' || 'KTX2'
```
Comment (ibgreen) [inline modules/textures/src/basis-loader.ts:22]: I recommend lower case strings for all option values. `'basis', 'ktx2'`

File: modules/textures/src/compressed-texture-loader.ts [hunk 4]
```diff
@@ -37,7 +40,13 @@ export const CompressedTextureWorkerLoader = {
  */
 export const CompressedTextureLoader = {
   ...CompressedTextureWorkerLoader,
-  parse: async (arrayBuffer) => parseCompressedTexture(arrayBuffer)
+  parse: async (arrayBuffer, options) => {
+    if (options['compressed-texture'].useBasis) {
+      options.basis = {...options.basis, decoderFormat: 'KTX2', module: 'encoder'};
```
Comment (ibgreen) [inline modules/textures/src/compressed-texture-loader.ts:45]: Can we use a more standard way of applying default options, see other loaders?

File: modules/textures/src/lib/parsers/parse-basis.js [hunk 5]
```diff
@@ -22,11 +22,30 @@ const OutputFormat = {
...
+      switch (options.basis.decoderFormat) {
+        case 'KTX2':
+          return parseKTX2File(fileConstructors.KTX2File, data, options);
+        case 'Basis':
+        default:
+          return parseBasisFile(fileConstructors.BasisFile, data, options);
+      }
+    case 'transcoder':
+    default:
+      const {BasisFile} = await loadBasisModule(options);
```
Comment (ibgreen) [inline modules/textures/src/lib/parsers/parse-basis.js:37]: Can we align naming of loadBasisModule and loadBasisEncoderModule?

File: modules/textures/src/lib/parsers/basis-module-loader.ts [hunk 6]
```diff
@@ -43,3 +43,56 @@ function initializeBasisModule(BasisModule, wasmBinary) {
...
+    )
+  ]);
+
+  // Depends on how import happened...
+  // @ts-ignore TS2339: Property does not exist on type
+  BASIS_ENCODER = BASIS_ENCODER || globalThis.BASIS;
+  return await initializeBasisEncoderModule(BASIS_ENCODER, wasmBinary);
+}
+
+function initializeBasisEncoderModule(BasisEncoderModule, wasmBinary) {
```
Comment (ibgreen) [inline modules/textures/src/lib/parsers/basis-module-loader.ts:83]: JSDoc on all functions. Mostly interested in a line that describes in word what the function does.

File: modules/textures/src/lib/parsers/parse-basis.js [hunk 7]
```diff
@@ -22,11 +22,30 @@ const OutputFormat = {
...
+      return parseBasisFile(BasisFile, data, options);
+  }
+}
+
+function parseBasisFile(BasisFile, data, options) {
   const basisFile = new BasisFile(new Uint8Array(data));
 
   try {
     if (!basisFile.startTranscoding()) {
+      basisFile.close();
```
Comment (ibgreen) [inline modules/textures/src/lib/parsers/parse-basis.js:47]: Maybe a `try {} finally {}`

**Review by ibgreen — APPROVED** [review]

File: modules/textures/src/compressed-texture-loader.ts (hunk 4 above)
Comment (belom88) [reply modules/textures/src/compressed-texture-loader.ts:45]: There could be values from user with custom `format`

File: modules/textures/package.json (hunk 2 above)
Comment (belom88) [reply modules/textures/package.json]: Done

File: modules/textures/src/basis-loader.ts (hunk 3 above)
Comment (belom88) [reply modules/textures/src/basis-loader.ts]: Done

File: modules/textures/src/lib/parsers/parse-basis.js (hunk 5 above)
Comment (belom88) [reply modules/textures/src/lib/parsers/parse-basis.js]: Done

File: modules/textures/src/lib/parsers/basis-module-loader.ts (hunk 6 above)
Comment (belom88) [reply modules/textures/src/lib/parsers/basis-module-loader.ts:115]: Done

File: modules/textures/src/lib/parsers/parse-basis.js (hunk 7 above)
Comment (belom88) [reply modules/textures/src/lib/parsers/parse-basis.js]: Done

**Pull request merged**